### PR TITLE
Add `DockerRegistry` block for `DockerContainer` to pull images from registries with authentication

### PR DESCRIPTION
**Pull request by madkinsz**
- Adds a `DockerRegistry` block with a `login` capability.
- Adds an `image_registry` setting to the `DockerContainer` infrastructure block allowing login to a registry before image pull.

Addresses https://github.com/PrefectHQ/prefect/issues/6230 — I do not think it is fair to close this since a separate implementation with ECR token refresh support will be required.

### CODE REVIEW

**Review by chrisguidry — APPROVED** [review]
I'm still surprised we have to do this at all, but reading through the related `docker-py` issue, it makes sense.  It also makes a lot of sense to be a block.

What would the username and password be for ECR?  The result of `aws ecr get-login-password ...`?

**Comment by madkinsz** [comment]
I'm not 100% sure on ECR, you can see some examples in the upstream issue https://github.com/docker/docker-py/issues/2256#issuecomment-887175827

**Comment by anna-geller** [comment]
This is awesome, thanks so much Michael! 

@chrisguidry yes correct, you would use this command to get the token which can be used as password, the username is always AWS:`docker login -u AWS -p your_token`

In theory, AWS considers this as insecure and recommends using --password-stdin:
```
aws ecr get-login-password --region region | docker login --username AWS --password-stdin aws_account_id.dkr.ecr.region.amazonaws.com
```

but either way, the token is only valid for 12 hours, so not sure whether after this initial login, dockerpy can consider just using permissions attached via IAM role to the EC2 instance...

**Comment by madkinsz** [comment]
We can add an `ElasticContainerRegistry` block that pulls the ECR token at runtime to the `prefect-aws` collection. cc @desertaxle

File: src/prefect/infrastructure/docker.py [hunk 1]
```diff
@@ -39,6 +40,63 @@ class ImagePullPolicy(AutoEnum):
...
+        password: The password to log into the registry with.
+        registry_url: The URL to the registry. Generally, "http" or "https" can be
+            omitted.
+        reauth: If already logged into the registry, should login be performed again?
+            This setting defaults to `True` to support common token authentication
+            patterns such as ECR.
+    """
+
+    _block_type_name = "Docker Registry"
+    _block_schema_capabilities = ["login"]
```
Comment (desertaxle) [inline src/prefect/infrastructure/docker.py:61]: I think the capability should be a little more specific
```suggestion
    _block_schema_capabilities = ["docker-login"]
```

File: src/prefect/infrastructure/docker.py [hunk 2]
```diff
@@ -39,6 +40,63 @@ class ImagePullPolicy(AutoEnum):
...
+    Attributes:
+        username: The username to log into the registry with.
+        password: The password to log into the registry with.
+        registry_url: The URL to the registry. Generally, "http" or "https" can be
+            omitted.
+        reauth: If already logged into the registry, should login be performed again?
+            This setting defaults to `True` to support common token authentication
+            patterns such as ECR.
+    """
+
```
Comment (desertaxle) [inline src/prefect/infrastructure/docker.py:59]: Here's a fun little Moby logo for this block
```suggestion
_logo_url = "https://images.ctfassets.net/gm98wzqotmnx/2IfXXfMq66mrzJBDFFCHTp/344dda583986d2d0db361c92dd650693/Moby-logo.webp?h=250"
```

File: src/prefect/infrastructure/docker.py (hunk 1 above)
Comment (madkinsz) [reply src/prefect/infrastructure/docker.py:61]: Agreed.

File: src/prefect/infrastructure/docker.py [hunk 3]
```diff
@@ -39,6 +40,63 @@ class ImagePullPolicy(AutoEnum):
...
+_logo_url = "https://images.ctfassets.net/gm98wzqotmnx/2IfXXfMq66mrzJBDFFCHTp/344dda583986d2d0db361c92dd650693/Moby-logo.webp?h=250"
+    _block_type_name = "Docker Registry"
+    _block_schema_capabilities = ["docker-login"]
+
+    username: str
+    password: SecretStr
+    registry_url: str
+    reauth: bool = True
+
+    def login(self):
```
Comment (madkinsz) [inline src/prefect/infrastructure/docker.py:68]: Might want this to be async first?

File: src/prefect/infrastructure/docker.py (hunk 2 above)
Comment (anna-geller) [reply src/prefect/infrastructure/docker.py:59]: @madkinsz the current logo doesn't render well, we can take the one from Alex for both this and DockerContainer Block 
image

**Comment by anna-geller** [comment]
This works 🎉 

image

image
 
QAed with S3 storage block and Dockerfile that does nothing but adding s3fs dependency:

```
FROM prefecthq/prefect:dev-python3.9
RUN pip install s3fs
```

**Review by anna-geller — APPROVED** [review]

**Comment by madkinsz** [comment]
@anna-geller This was waiting for a few things still :) I'll open a follow-up.

**Comment by feliperazeek** [comment]
@anna-geller I am going over your docker-based [sample project](https://github.com/anna-geller/prefect-docker-deployment/), how would I modify it to use ECR? thank you very much!

**Comment by zanieb** [comment]
@feliperazeek I'd recommend opening a new issue instead of bumping this old pull request.

**Pull request merged**